src/python/gudhi/tensorflow/cubical_layer.py

```diff
@@ -0,0 +1,66 @@
+import numpy               as np
+import tensorflow          as tf
+from ..cubical_complex  import CubicalComplex
+
+######################
+# Cubical filtration #
+######################
+
+# The parameters of the model are the pixel values.
+
+def _Cubical(X, dimension):
+    # Parameters: X (image),
+    #             dimension (homology dimension)
+
+    # Compute the persistence pairs with Gudhi
+    cc = CubicalComplex(dimensions=X.shape, top_dimensional_cells=X.flatten())
+    cc.persistence()
+    try:
+        cof = cc.cofaces_of_persistence_pairs()[0][dimension]
+    except IndexError:
+        cof = np.array([])
+
+    if len(cof) > 0:
+        # Sort points with distance-to-diagonal
+        Xs = X.shape
+        pers = [X[np.unravel_index(cof[idx,1], Xs)] - X[np.unravel_index(cof[idx,0], Xs)] for idx in range(len(cof))]
+        perm = np.argsort(pers)
+        cof = cof[perm[::-1]]
+    
+    # Retrieve and ouput image indices/pixels corresponding to positive and negative simplices
+    D = len(Xs) if len(cof) > 0 else 1
+    ocof = np.array([0 for _ in range(D*2*cof.shape[0])])
+    count = 0
+    for idx in range(0,2*cof.shape[0],2):
+        ocof[D*idx:D*(idx+1)]     = np.unravel_index(cof[count,0], Xs)
+        ocof[D*(idx+1):D*(idx+2)] = np.unravel_index(cof[count,1], Xs)
+        count += 1
+    return np.array(ocof, dtype=np.int32)
+
+class CubicalLayer(tf.keras.layers.Layer):
+    """
+    TensorFlow layer for computing cubical persistence out of a cubical complex
+
+    Attributes:
+        dimension (int): homology dimension
```

Comment:
So there is one fixed dimension. What is the plan if my loss depends on the diagrams in dimensions 0 and 1? Should I make several independent layers, and hope :crossed_fingers: that the cost of the highest dimension dominates so it isn't too bad?

Reply:
That's more or less the plan for now... but indeed this feels really suboptimal

Reply:
At least with this interface it remains possible to add later the possibility to pass something like `dimension=(1,2)` to handle multiple dimensions. I think it is ok for a tensorflow layer to take or return a tuple of tensors instead of just one tensor.

Reply:
Yep, you are right, it was quite easy to do so I added the possbility to ask for multiple dimensions at once.

Reply:
I thought it would be: if we pass an int, return the diagram in that dim, if we pass a tuple/list of dimensions, return a tuple/list of diagrams, but maybe that's overkill, always using tuple/list is indeed probably sufficient, and it is simpler :smiley: